'''
Main Snakefile.
The aim is to run the cellranger pipeline automatically on all the samples defined in the config file
'''

# Path of the config file
configfile: 'config/config.yaml'

# Rules to execute the workflow
include: 'rules/cellranger_processing.smk'
include: 'rules/cellranger_IntronExon.smk'



# Master rules to launches the workflow
rule minimal_cellranger:
    '''
    rule to run the default analysis using cellranger.
    '''
    input:
        expand(rules.runCellranger.output.output, sample=config['SAMPLES'])


# default rule to run the cellranger pipeline
rule default_cellranger:
    '''
    rule to run the default analysis using cellranger.
    '''
    input:
        expand(rules.moveCellrangerSummary.output.summary, sample=config['SAMPLES']),
        rules.runMultiqc1.output.html

# rule to lauch default cellranger with the option to use the intronic exonic pure fitlering
rule pure_IntronExon:
    '''
    rule to produce the pure exon bam file.
    '''
    input:
        expand(rules.moveCellrangerSummary.output.summary, sample=config['SAMPLES']),
        expand(rules.moveCellrangerIntronExon.output.bam_file, sample=config['SAMPLES']),
        expand(rules.runPureExon.output.exon_count, sample=config['SAMPLES']),
        expand(rules.runPureIntron.output.intron_count, sample=config['SAMPLES']),
        expand(rules.runPureAll.output.all_count, sample=config['SAMPLES']),
        rules.runMultiqc1.output.html,
        rules.copyGtf.output.gtf

# rule all_cellranger_test:
#     '''
#     rule to run the default analysis using cellranger.
#     '''
#     input:
#         expand(rules.runCellranger_alt.output.output, sample=config['SAMPLES'])

# ---------------------------------------------------------------------------- #
# Test rule to make sure the Snakemake setup is working
rule all_test:
    '''
    testing rule
    '''
    input:
        rules.test.output.test
